src/app.js

```diff
@@ -215,48 +217,134 @@ export default function (app) {
-
-      if (oldNode && oldNode.data && oldNode.data.onRemove) {
-        defer(oldNode.data.onRemove, element)
+      if (node.data && node.data.onRemove) {
+        defer(node.data.onRemove, element)
       }
+  }
+
+  function patch (parent, element, oldNode, newNode) {
+    if (oldNode === undefined) {
+      var n = createElementFrom(newNode)
+      parent.insertBefore(n, element)
+      element = n
 
     } else if (
-      node.tag !== oldNode.tag ||
-      typeof node !== typeof oldNode ||
-      typeof node === "string" && node !== oldNode
+      newNode.tag !== oldNode.tag ||
+      typeof newNode !== typeof oldNode ||
+      typeof newNode === "string" && newNode !== oldNode
     ) {
-      if (typeof node === "string") {
-        element.textContent = node
-      } else {
-        var i = createElementFrom(node)
-        parent.replaceChild(i, element)
-        element = i
-      }
-    } else if (node.tag) {
-      updateElementData(element, node.data, oldNode.data)
+      var n = createElementFrom(newNode)
+      parent.replaceChild(n, element)
+      element = n
+
+    } else if (oldNode.tag){
+      updateElementData(element, newNode.data, oldNode.data)
+      patchChildren(element, oldNode, newNode)
+    }
+
+    return element
+  }
+
+  function patchChildren(element, oldNode, newNode) {
+    // We need to index all keyed nodes in oldChildren in order to
+    // distinguish new keyed nodes from nodes to reuse in newChildren
+    var keyedChildren = {}
+    for (var i = 0; i < oldNode.children.length; i++) {
+      if (!oldNode.children[i].data) continue
+      if (!oldNode.children[i].data.key) continue
+      keyedChildren[oldNode.children[i].data.key] = true
+    }
 
-      var len = node.children.length
-      var oldLen = oldNode.children.length
+    var isKeyed = function (node) {
+      return (
+        !!node.data &&
+        !!node.data.key &&
+        !!keyedChildren[node.data.key]
+      )
+    }
+
+    var newIndex = 0
+    var oldIndex = 0
+    var indexOffset = 0
+    var cache = {}
+    var oldChild
+    var newChild
+
+    var patchChild = function () {
+      patch(element, element.childNodes[oldIndex + indexOffset], oldChild, newChild)
+      oldIndex++
+      newIndex++
+    }
+
+    var insertChild = function () {
+      patch(element, element.childNodes[oldIndex + indexOffset], undefined, newChild)
+      indexOffset++
+      newIndex++
+    }
+
+    var removeChild = function () {
+      batchRemove(element, element.childNodes[oldIndex + indexOffset], oldChild)
+      oldIndex++
+    }
+
+    while (
+      newNode.children[newIndex] !== undefined ||
+      oldNode.children[oldIndex] !== undefined
+    ) {
+      newChild = newNode.children[newIndex]
+      oldChild = oldNode.children[oldIndex]
+
+      if (
+        newChild !== undefined &&
+        isKeyed(newChild) &&
+        cache[newChild.data.key]
+      ) {
+        var c = cache[newChild.data.key]
```

Comment:
You mean `var i` compresses better than `var c`? Is that because `var i` is already used in a lot if places? 

Either way, I'll fix that, no problem.